design_docs/rsyslog_new_input_output_roles_proposal.md

```diff
@@ -0,0 +1,255 @@
+- how to support more generic way to connect input roles and output roles.
+
+Please note that this github commit starting this doc implements the first cut of the proposal including:
+- input roles - basics input role (`imjournal`, `imuxsock`, `imudp`, `imptcp`) and files input role (`imfile`), and
+- output roles - files output role (`omfile`) and forwards output role (`omfwd`).
+Leftover tasks, other roles to be changed and eliminating `rsyslog_default`, are discussed in [To Do's](#to-dos).
+
+## Proposals
+
+### Output roles
+Each output role would be stored in a configuration file which suffix is a type name.
+
+Files and Forwards output configuration files in /etc/rsyslog.d:
+```
+  10-output-modules.conf
+  30-output.files
+  30-output.forwards
+```
+
+### Ruleset to switch the output
+Using rsyslog rulesets, we prepare all possible output combinations in the ruleset configuration file as follows.
+Note: The ruleset name is made from the output types in the alphabetical order.
+```
+==> 50-default-rulesets.conf <==
+# Ansible managed
+
+#
+# Rules for the local system logs
+#
+ruleset(name="files") {
+  $IncludeConfig /etc/rsyslog.d/*.files
+}
+ruleset(name="forwards") {
+  $IncludeConfig /etc/rsyslog.d/*.forwards
+}
+```
+In this PR, just files and forwards outputs are supported.  As described in [To Do's](#to-dos), elasticsearch is going to be added next.
+
+### Input roles
+Based upon the inventory file that users provide, input config jumps to the ruleset.
+
+The next `imjournal` (basics) example shows log messages read from journald by the `imjournal` plugin are logged in the local files.
+
+**`imjournal` example**
+```
+==> 10-basics-modules.conf <==
+# Ansible managed
+
+#
+# System log messages
+#
+# Read from journald
+module(load="imjournal"
+       StateFile="/var/lib/rsyslog/imjournal.state"
+       RateLimit.Burst="20000"
+       RateLimit.Interval="600"
+       PersistStateInterval="10")
+call files
+stop
+```
+
+**`imfile` example**
+The next `imfile` example shows log messages read from /var/log/inputdirectory/*.log by the `imfile` plugin are logged in the local files as well as forwarded to the remote host.
+```
+==> 90-imfile-input.conf <==
+# Ansible managed
+
+#
+# Log messages from log files
+#
+
+input(type="imfile" file="/var/log/inputdirectory/*.log" tag="container")
+call files
+call forwards
+stop
+```
+
+### Inventory file changes
+Here is an example that the inputs from `imjournal` (basics) are logged in the local file;
+the inputs from `imfile` (files) are forwarded to the remote host in addition to the local file,
+
+**Proposed style**
+```
+logging_outputs:
+  - name: forward-output
+    type: forwards
+    rsyslog_forwards_actions:
+      - name: to-remote
+        protocol: tcp
+        target: remote_host_name.remote_domain_name
+        port: 514
+  - name: file-output
+    type: files
+logs_collections:
```

Comment:
should be `logging_collections`, no? According to standards. Also, wouldn't it be more intuitive to call it `inputs` for symmetry with `outputs`, instead of `collections`?

Reply:
I like the idea.
@sradco, is it ok to change the variable name?

Reply:
@nhosoi Yes.it makes sense. I'll need to update this for ovirt.

Reply:
Thanks, @sradco!
@pcahyna, is it ok to make the change in this PR or do you recommend to create a separate PR?

Reply:
if you mean the change to the `logs_collections` -> `logging_inputs` variable name, then please do it in this PR. It would be just confusing to introduce an unwanted variable name in one PR and then change it in another one.

Reply:
??? logs_collections has been in the logging role for a long time.  So, it's just a replacement.  If you don't mind, I'm more than happy to do it in this PR.

Reply:
I don't see it. I currently see `log_collections` only as an attribute of dicts. With the proposed change, you would turn it into a variable. I.e. introduce a new variable.

Reply:
Thank you for finding out the typo... :)  I pushed the patch with the typo fix.
https://github.com/linux-system-roles/logging/pull/100/commits/ca6eb843d284774df75a79bb1cdab2e0958083ce

Reply:
??? I did not find any typo, I just explained my reasoning for saying that `logs_collections` was a newly introduced variable in the PR.